src/metaxy/metadata_store/base.py

```diff
@@ -480,18 +489,65 @@ def write_metadata(
         if not is_system_table:
             self._validate_project_write(feature)
 
-        # Convert Narwhals to Polars if needed
-        if isinstance(df, nw.DataFrame):
-            df = df.to_polars()
-        # nw.DataFrame also matches as DataFrame in some contexts, ensure it's Polars
-        if not isinstance(df, pl.DataFrame):
-            # Must be some other type - shouldn't happen but handle defensively
-            if hasattr(df, "to_polars"):
-                df = df.to_polars()
-            elif hasattr(df, "to_pandas"):
-                df = pl.from_pandas(df.to_pandas())
-            else:
-                raise TypeError(f"Cannot convert {type(df)} to Polars DataFrame")
+        # Convert Narwhals to native Polars (both lazy and eager)
+        if isinstance(df, nw.LazyFrame):
+            df = df.to_native()  # type: ignore[assignment]
+        elif isinstance(df, nw.DataFrame):
+            df = df.to_native()  # type: ignore[assignment]
+
+        if pa is not None and isinstance(df, pa.Table):
+            # Streaming paths may yield Arrow tables; normalize to Polars.
+            df = cast(pl.DataFrame, pl.from_arrow(df))
+
+        # Now df is either pl.DataFrame or pl.LazyFrame
+
+        # Collect lazy frames unless subclass wants to handle them
+        if self._auto_collect_lazy_frames and isinstance(df, pl.LazyFrame):
```

Comment:
I don't really like this pattern in the base store. 

We **never** collect DataFrames in the whole base store code path. There is only one exception, and it emits a warning, and we make sure it never happens in tests unless explicitly slowed: when falling back to the Polars provenance tracker from the native one. 

There is basically no reason to perform this collect, am I wrong here? Even if we decide to introduce this option, I would default it to False - that's just safer.